riscv/sim.cc

```diff
@@ -43,13 +42,11 @@ sim_t::sim_t(const char* isa_string, const char* priv, const char* varch,
 #endif
              FILE *cmd_file) // needed for command line option --cmd
   : htif_t(args),
-    isa(isa_string, priv),
+    isa(base_cfg->get_isa().get(), base_cfg->get_priv().get()),
+    base_cfg(base_cfg),
     mems(mems),
     plugin_devices(plugin_devices),
-    procs(std::max(nprocs, size_t(1))),
-    initrd_start(initrd_start),
-    initrd_end(initrd_end),
-    bootargs(bootargs),
+    procs(std::max(base_cfg->get_nprocs().get(), size_t(1))),
```

Comment:
Hmm, I think I'd quite like to keep the setters because they let you do things like (minor) format conversions on the fly. For example, the initrd range stuff. But, obviously, that doesn't work well with exposing the members as public attributes.

How about supplying `operator->` instead of `operator()`? The result would look a bit like a smart pointer, and you'd have something like `*base_cfg->get_nprocs()`. Still a bit of punctuation but it's not as ridiculous as what I had.

(Happy to implement this, but I thought I'd get feedback before rehashing all the code!)

Reply:
Why not just have the callers use `std::pair` directly for initrd? A class with hardly anything but getters & setters doesn't have a good smell.

I'm not really following how the `operator->` would work...somebody has to call `nprocs.get()`, so would that mean that `operator->` would return a new class with a bunch of getter methods?

Reply:
Well, I was imagining that we kept the getter, so `get_nprocs()` returns a `cfg_arg_t<int>`, which has an `operator->` to extract the `int`. This would be analogous to the way a `std::unique_ptr<T>` has `get()` and `operator->` that do the same thing.

I'm happy to drop the setters and getters entirely. I don't think that the "access control" argument is all that strong, so that seems quite a reasonable thing to do. If we do that (and make no other changes), we get `cfg.nprocs.get()`. I'm not sure about adding the `operator()`: you end up with `cfg.nprocs()`, but that looks like we're calling a method called `nprocs` on `cfg`, which seems a bit confusing.

Maybe `cfg.nprocs.get()` is terse enough? Or we could do the `operator->` thing to end up with `*cfg.nprocs`: terser, but maybe more magical.

What do you think?

Reply:
Cool, sounds good to me! I've just pushed up a version that does this.